Build failure: haskell.compiler.ghcjs

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

Using the below flake, `ghcjs` builds successfully (albeit after about an hour):

```nix
{
  description = "GHCJS Flake";

  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-24.11";
  };

  outputs =
    {
      self,
      nixpkgs,
      flake-utils,
    }:
    flake-utils.lib.eachDefaultSystem (
      system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
      in
      {
        devShells.default = pkgs.mkShell {
          packages = with pkgs; [
            haskell.compiler.ghcjs
          ];
        };
      }
    );
}
```

However, when switching to unstable, it fails to build:

```nix
{
  description = "GHCJS Flake";

  inputs = {
    flake-utils.url = "github:numtide/flake-utils";
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
  };

  outputs =
    {
      self,
      nixpkgs,
      flake-utils,
    }:
    flake-utils.lib.eachDefaultSystem (
      system:
      let
        pkgs = nixpkgs.legacyPackages.${system};
      in
      {
        devShells.default = pkgs.mkShell {
          packages = with pkgs; [
            haskell.compiler.ghcjs
          ];
        };
      }
    );
}
```

### Can Hydra reproduce this build failure?

No, Hydra cannot reproduce this build failure.

### Link to Hydra build job

_No response_

### Relevant log output

```console
error: builder for '/nix/store/nn0fwrg4r2y3qhj9m8jhaypm4hy5g8ja-ghcjs-8.10.7.drv' failed with exit code 1;
       last 25 log lines:
       > /build/configured-ghcjs-src/lib/boot/pkg/integer-gmp$ /nix/store/81j9lximsskg386b0qya6dn3g1dnkrdm-cabal-install-3.14.1.0/bin/cabal "v1-configure" "--global" "--ghcjs" "--builddir" "dist" "--with-compiler" "/nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/bin/ghcjs" "--with-hc-pkg" "/nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/bin/ghcjs-pkg" "--with-haddock" "/nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/bin/haddock" "--with-gcc" "/nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/lib/ghcjs-8.10.7/bin/emcc" "--prefix" "/nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7" "--configure-option" "--host=js-unknown-ghcjs" "--ghcjs-options=-fwrite-ide-info" "--enable-debug-info" "--disable-library-stripping" "--disable-executable-stripping" "-fghci" "--enable-profiling" "--enable-shared"
       > Warning: Unknown/unsupported 'ghc' version detected (Cabal 3.14.1.0 supports
       > 'ghc' version < 8.8):
       > /nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/bin/ghcjs is based on
       > GHC version 8.10.7
       > Warning: No remote package servers have been specified. Usually you would have
       > one specified in the config file.
       > Resolving dependencies...
       > Warning: solver failed to find a solution:
       > Could not resolve dependencies:
       > [__0] next goal: integer-gmp (user goal)
       > [__0] rejecting: integer-gmp-1.0.3.0 (constraint from non-reinstallable
       > package requires installed instance)
       > [__0] fail (backjumping, conflict set: integer-gmp)
       > After searching the rest of the dependency tree exhaustively, these were the
       > goals I've had most trouble fulfilling: integer-gmp
       > Trying configure anyway.
       > Configuring integer-gmp-1.0.3.0...
       > configure: WARNING: unrecognized options: --with-compiler, --with-hc-pkg
       > checking for js-ghcjs-gcc... /nix/store/8psfzn28z08wwlnwbr1fl8p3m3rbrfgl-ghcjs-8.10.7/lib/ghcjs-8.10.7/bin/emcc
       > checking whether the C compiler works... no
       > configure: error: in '/build/configured-ghcjs-src/lib/boot/pkg/integer-gmp/dist/build':
       > configure: error: C compiler cannot create executables
       > See 'config.log' for more details
       > fatal: program /nix/store/81j9lximsskg386b0qya6dn3g1dnkrdm-cabal-install-3.14.1.0/bin/cabal returned a nonzero exit code
       For full logs, run 'nix log /nix/store/nn0fwrg4r2y3qhj9m8jhaypm4hy5g8ja-ghcjs-8.10.7.drv'.
error: 1 dependencies of derivation '/nix/store/hqz8vyr6sp5s4nj8mwq0ibh0gnns7ldk-nix-shell-env.drv' failed to build
```

### Additional context

(note, Hydra cannot build this as it produces artifacts too large)

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.10, NixOS, 25.05 (Warbler), 25.05.20250121.9e4d519`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - nixpkgs: `/nix/store/irknf7niks2djxz4csh7sr44wwrs082y-source`

### Notify maintainers

@obsidian-systems-maintenance 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/compilers/ghcjs/8.10/default.nix
+++ b/pkgs/development/compilers/ghcjs/8.10/default.nix
@@ -105,6 +105,9 @@ stdenv.mkDerivation {
     touch $HOME/.cabal/config
     cd lib/boot
 
+    export EM_CACHE="$HOME/.emscriptencache"
+    mkdir -p "$EM_CACHE"
+
     mkdir -p $out/bin
     mkdir -p $out/lib/${bootGhcjs.name}
     lndir ${bootGhcjs}/bin $out/bin

